examples/noise-js/index.js

```diff
@@ -0,0 +1,357 @@
+    // contains private information that needs to be stored safely
+    const contentTopic = pairingObj.contentTopic;
+    const handshakeResult = pairingObj.getHandshakeResult();
+
+    // To restore the codecs for decrypting older messages, or continuing an existing
+    // session, use this:
+    [encoder, decoder] = WakuPairing.getSecureCodec(contentTopic, handshakeResult);
+    */
+    ui.message.display();
+
+    await node.filter.subscribe(
+      [decoder],
+      ui.message.onReceive.bind(ui.message)
+    );
+
+    ui.message.onSend(async (text, nick) => {
+      const timestamp = Math.floor(Date.now() / 1000);
+      const message = ProtoChatMessage.create({
+        nick,
+        timestamp,
+        text: utils.utf8ToBytes(text),
+      });
+      const payload = ProtoChatMessage.encode(message).finish();
+
+      await node.lightPush.push(encoder, { payload, timestamp });
+    });
+  } catch (err) {
+    ui.waku.error(err.message);
+    ui.hide();
+  }
+}
+
+function buildPairingURLFromObj(pairingObj) {
+  const pInfo = pairingObj.getPairingInfo();
+
+  // Data to encode in the QR code. The qrMessageNametag too to the QR string (separated by )
+  const messageNameTagParam = `messageNameTag=${utils.bytesToHex(
+    pInfo.qrMessageNameTag
+  )}`;
+  const qrCodeParam = `qrCode=${encodeURIComponent(pInfo.qrCode)}`;
+
+  return `${window.location.href}?${messageNameTagParam}&${qrCodeParam}`;
+}
+
+function getPairingInfoFromURL() {
+  const urlParams = new URLSearchParams(window.location.search);
+
+  const messageNameTag = urlParams.get("messageNameTag");
+  const qrCodeString = urlParams.get("qrCode");
+
+  if (!(messageNameTag && qrCodeString)) {
+    return undefined;
+  }
+
+  return new noise.InitiatorParameters(
+    decodeURIComponent(qrCodeString),
+    utils.hexToBytes(messageNameTag)
+  );
+}
+
+function getSenderAndResponder(node) {
+  const sender = {
+    async publish(encoder, msg) {
+      await node.lightPush.push(encoder, msg);
+    },
+  };
+
+  const msgQueue = new Array();
+  const subscriptions = new Map();
+  const intervals = new Map();
+
+  const responder = {
+    async subscribe(decoder) {
+      const subscription = await node.filter.subscribe(
+        [decoder],
+        (wakuMessage) => {
+          msgQueue.push(wakuMessage);
+        }
+      );
+      subscriptions.set(decoder.contentTopic, subscription);
+    },
+    async nextMessage(contentTopic) {
+      if (msgQueue.length != 0) {
+        const oldestMsg = msgQueue.shift();
+        if (oldestMsg.contentTopic === contentTopic) {
+          return oldestMsg;
+        }
+      }
+
+      return new Promise((resolve) => {
+        const interval = setInterval(() => {
+          if (msgQueue.length != 0) {
+            clearInterval(interval);
+            const oldestMsg = msgQueue.shift();
+            if (oldestMsg.contentTopic === contentTopic) {
+              resolve(oldestMsg);
+            }
+          }
+        }, 100);
```

Comment:
Seems like `WakuPairing` needs this interface passed.

But it is strange that the developer need to construct a `sender` and `responder` that seems to just be an arbtirary usage of the filter and lightpush protocols.

I suggest we track some improvement here on a separate issue.

I am guessing the issue is that the noise pairing object needs to have a way to send and receive messages, and it's up to the user to decide what protocols (light push, filter, relay) should be used.

Something I already thought about in the past and now would be a good reason to do it, is to create a `Sender` and `Receiver` interface in `@waku/interfaces` and have:
- lightpush and relay implements `Sender`
- filter and relay implements `Receiver`

then, the user can pass `node.lightpush` and `node.filter` to `WakuPairing` without these 50 lines of code.

Reply:
To be addressed in this issue https://github.com/waku-org/js-noise/issues/14